revise mirage_net interface

hannesm opened the pull request:
- ~~remove unused `writev` (unsure about this, nobody called it though)~~
- provide `mtu : t -> int`
- provide `allocate_frame : t -> buffer`
- remove `type page_aligned_buffer`

//cc @mirage/core 
The new model in respect to allocation of network frames: let the lowest layer do it, it knows best which MTU and alignment characteristics it needs. On the way we can remove `io_page` from higher layers - but a `mirage-net-xen` will hand out properly page-aligned frames - unfortunately I don't know of a good way to ensure whatever buffer `write` gets was previously allocated by `allocate_frame` (while still allowing arbitrary Cstruct.t operations on the buffer). I plan to add `allocate_frame` to each layer in the stack: TCP calls `IP.allocate_frame`, which calls `Ethernet.allocate_frame`, which calls `Netif.allocate_frame`, which does the allocation. On the way back the offset is increased for each layer, The `write` function in each layer encodes its header into the frame and calls `write` on the lower layer.  

PRs
- [ ] provider https://github.com/mirage/mirage-net-solo5/pull/26
- [ ] provider https://github.com/mirage/mirage-vnetif/pull/24
- [ ] provider https://github.com/mirage/mirage-net-unix/pull/44
- [ ] provider https://github.com/mirage/mirage-net-macosx/pull/28
- [ ] consumer (test) https://github.com/mirage/mirage-tcpip/pull/379
- [ ] consumer https://github.com/mirage/ethernet/pull/2
- [ ] provider https://github.com/mirage/mirage-net-xen/pull/78

hannesm committed:
revise mirage_net interface:
- provide mtu : t -> int
- provide allocate_frame : t -> buffer
- remove page_aligned_buffer

avsm commented on src/mirage_net.mli:
```diff
@@ -42,42 +42,47 @@ type stats = {
+  (** [allocate_frame net] allocates a single frame which can be filled
+      and transmitted via [write]. The network interface implementation knows
+      best about its memory-alignment requirements. *)
+
```
Lgtm. Perhaps add a note here at that mixing frames allocated between different network devices is currently undefined, so dont do it?

hannesm replied:
yes, I'll do that.

hannesm commented:
another question while thinking more about it: should the signature rather be `allocate_frame : ?requested_size:int -> t -> buffer`? This would allow the upper layers to specify a maximum value, and avoid expensive allocation if the mtu is extraordinary large (10k, 65k or similar, i.e. support for jumbo frames)! WDYT?
EDIT: I don't like `requested_size`, if someone has an idea for a name, I'm interested :)

hannesm commented:
some notes:
- (a) the mtu is not discovered in ocaml-tuntap yet (there's the `ioctl` `TAPSIFINFO` available, according to my `man tap`, but I don't know how other POSIX systems do this, anyone interested in figuring out and updating both tuntap and mirage-net-unix
- (b) on Xen I've no idea how to get the mtu, and while looking into mirage-net-xen I'm very unfamiliar with the code, anyone up for adjusting it to the interface changes proposed here?
- (c) what is the status of mirage-net-fd and mirage-net-flow? are these libraries actively used and need to be adapted as well? if so, any volunteers?

avsm commented:
Requested_size seems fine to me -- the buffer actually allocated may be larger than what is asked for (if there are page size constraints), or smaller (if there are not enough contiguous pages available for dma buffers), so a request seems like a good phrase here.

a) it may be hard to do this portably :-/ if you could open an issue i could take a look, but it might just be better to hardcode some sensible mtu for this setup like 1500

b) xen it is available by looking if the `sg` feature flag is set in xenstore; if so it is 0xffff (see [linux netfront](https://elixir.bootlin.com/linux/v4.9.12/source/include/xen/interface/io/netif.h#L834)) or otherwise its smaller (need to check the netfront driver).
@jonludlam may know more...

c) i'm not sure -- @djs55, did you do mirage-net-fd?

hannesm committed:
size argument for allocate_frame

hannesm commented:
(0) I added a `?size:int` argument to allocate_frame (and updated the PRs)
(a) that is what is currently in mirage-net-unix (1514 to be precise)
(b) would be great if someone does a PR to mirage-net-xen with that
(c) @samoht wrote them according to github

hannesm committed:
no need for io-page anymore

hannesm commented:
hello, to move forward I developed a PR for mirage-net-xen to comply with this revised interface. This as well contains a hardcoded mtu since I couldn't figure out how to get it from xen -- maybe @jonludlam @djs55 @avsm can take a look at https://github.com/mirage/mirage-net-xen/pull/78

I'd like to merge this rather sooner than later, any further comments to this? /cc @mirage/core

hannesm committed:
adjust constraints

hannesm committed:
pin to 2.0.0 temporarily

hannesm commented:
i rebased above mentioned PRs, and setup a mirage-dev branch where CI succeeds (see https://travis-ci.org/mirage/mirage-dev/builds/480597537?utm_source=github_status&utm_medium=notification - apart from the genode failure (I never got travis to build genode+mirage :/))

@mirage/core I would appreciate any opinion/work on:
(a) mirage-net-fd
(b) mirage-net-flow
(c) someone review and test the mirage-xen PR with this universe (don't have a xen machine)
(d) someone fix, review, and test the mirage-net-macosx PR above

merge + release plan: this weekend :)

hannesm commented:
an update on that: 
- (a) and (b) (mirage-net-flow/mirage-net-fd) don't seem to be used, i talked with dave scott who said they're not used in vpnkit.
- the proposed new interface is a bit rubbish (when it comes to lifetime guarantees etc.), modeled after the ip interface of `allocate_frame` and `write`. instead, i developed an interface (and pushed it across layers) that contains `write : t -> size:int -> (buffer -> unit) -> (unit, error) result io` -- with this, the network interface can allocate the buffer, and the provided function fills the buffer (i.e. lifetime of the buffer is strictly limited). i'm in the process to acquire some benchmarks with this new interface and fix up more packages with the new one.
- turns out the MTU from a tap device / ethernet interface is 1500 (i.e. MTU without ethernet header), so the implementation has to deal with this
- turns out, tuntap etc. can easily provide the mtu information (I've some partially working patches somewhere, will PR soon)

once I'm satisfied with the revised API and benchmarks, I'll followup on mirageos-devel for discussion.

samoht commented:
Indeed don't block on the `(a)` and `(b)`: my goal was to use them in [LinuxKit's Mirage SDK](https://github.com/linuxkit/linuxkit/tree/master/projects/miragesdk) to directly read packets from a raw device interface (using `ocaml-rawlink`) and to be able to do magic fd passing with a raw socket. but I haven't integrated that upstream properly yet. Ideally, we would just provide a `mirage-net-raw` package and extend the Unix backend to use that new network stack. But well, there is no direct use for this yet :-)

samoht commented on src/mirage_net.mli:
```diff
@@ -42,42 +43,50 @@ type stats = {
 
   val listen: t -> (buffer -> unit io) -> (unit, error) result io
-  (** [listen nf fn] is a blocking operation that calls [fn buf] with
+  (** [listen net fn] is a blocking operation that calls [fn buf] with
```
Might be interesting to specify here if the `buffer` is supposed to be own by the network interface and might be re-used when the callback is finished or if the consumer can do whatever it wants with the buffer (for instance store it in an external table) because it was copied before calling the callback.

Something like "the lifetime of the buffer is expected to be the managed by the network device, if you want to use the read buffer safely outside of the callback, copy the buffer manually." What do you think ?

hannesm replied:
yes - it's about lifetime and ownership. AFAICT `listen` (i looked into solo5 and unix implementations) actually (a) allocates a buffer (b) reads from the network device (c) creates an asynchronous task and passes the buffer to it. lifetime thereof is until the task finished and ownership is transferred to the task.

samoht replied:
So reading packets means allocating a lots of page ? This seems sub-optimal. Would probably be better if unix and solo5 implementation were using a pool of pages (and block/discard if no pages are available ?). Or should the buffer allocation function be passed as argument as well ?

hannesm replied:
hmm, I spent most of my recent time on the sending path, and figured lifetime/ownership/allocation there. on the receiving path, I'm not yet sure what a sensible way to deal would be:
(a) very simple solution: `listen` allocates initially a single buffer, and does not create any tasks, but instead calls `fn` after it read from the interface in the same task -- only once that `fn` is finished, `read` is called again, reusing the very same page. needs to state that `fn` needs to copy if it wants to retain the data
(b) `listen` uses a buffer pool - pre-allocating N pages initially, then calling read with the next free one, and creating an asynchronous task for each `fn` invocation -- but how should `fn` communicate that the given buffer is not used anymore? xen seems to use a ringbuffer and id interface, where the guest pushes memory into the ringbuffer with identifiers, and the host reports which identifiers are sent and can be safely reused/deallocated..
(c) the `fn` passed to `listen` could be modified to `buffer -> unit`, i.e. not in the io monad. this would allow an (a) solution (but still does not avoid the issue of storing a reference to the buffer anywhere).

my experience from the sending side is that the network device should be responsible for allocating (and it is the one which knows what buffer to pass to the underlying device, i.e. some need it page-aligned, others don't care), plus it knows the MTU. I don't think upper layers should provide allocators downwards.

but somehow the re-using of buffers scares me, we can't verify that any `fn` does not hold a reference to that buffer and reads it at a later stage - which may as well be a security concerns since now your web server can read the buffer used for receiving syslog messages. I'd propose to think a bit more about that, implement several strategies and evaluate the performance. I'll hopefully get around to writeup and push my sending part modifications this weekend.

samoht replied:
I'm fine to think / improve this later ; all your nice changes in the sending side had make me think on the receiving one :-) But no hurry and this don't need to be tied to that PR at all and can be done later.

hannesm commented:
superseeded by #16, thanks for your input!
